### PR TITLE
[4.0] [plg_privacy_user] Convert to prepared statement

### PR DESCRIPTION
**Pull request by HLeithner**
# Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.

### Testing Instructions
Use the plugin in all ways you can think of.

Create users
Delete user
Export privacy reports

### Expected result
Nothing changed.

### CODE REVIEW

**Comment by brianteeman** [comment]
2 unrelated files

File: plugins/privacy/user/user.php [hunk 1]
```diff
@@ -107,6 +108,8 @@ public function onPrivacyRemoveData(PrivacyTableRequest $request, JUser $user =
 			return;
 		}
 
+		$db = $this->db;
+
 		$pseudoanonymisedData = array(
```
Comment (Quy) [inline plugins/privacy/user/user.php]: Convert to short array syntax.

File: .drone.yml [hunk 2]
```diff
@@ -109,11 +109,53 @@ steps:
     commands:
       - bash tests/Codeception/drone-system-run.sh "$(pwd)" mysql
 
+  - name: artifacts-system-mysql
```
Comment (wilsonge) [inline .drone.yml]: Revert this please :)

Comment (HLeithner) [reply .drone.yml]: I mixed up the 3 PRs I did last night... changing drone in a prepared statement PR is not planed ;-)

File: plugins/privacy/user/user.php [hunk 3]
```diff
@@ -134,20 +140,21 @@ public function onPrivacyRemoveData(PrivacyTableRequest $request, JUser $user =
...
 		}
 
-		$this->db->setQuery(
-			$this->db->getQuery(true)
-				->delete($this->db->quoteName('#__session'))
-				->where($this->db->quoteName('session_id') . ' IN (' . implode(', ', $quotedIds) . ')')
-		)->execute();
+		$query->clear()
+			->delete($db->quoteName('#__session'))
+			->whereIn($db->quoteName('session_id'), $ids);
```
Comment (SharkyKZ) [inline plugins/privacy/user/user.php]: `session_id` column has `varbinary` type in MySQL and `bytea` in PgSQL. Won't this be a problem?

Comment (SharkyKZ) [reply plugins/privacy/user/user.php]: Seems like not. But should at least use `ParameterType::STRING` here, I think.

Comment (HLeithner) [reply plugins/privacy/user/user.php]: We have this pr for 3.9 https://github.com/joomla/joomla-cms/pull/23213 which  uses a new function quoteBinary I'm not sure how to do this with prepared statements. 

/cc @alikon can you help out here?

Comment (alikon) [reply plugins/privacy/user/user.php]: I have a vague  remembrance of the of use `PDO::PARAM_LOB`  in these cases, but not sure, need some tests to confirm

Comment (alikon) [reply plugins/privacy/user/user.php]: an usefull reading on this https://evertpot.com/writing-sql-for-postgres-mysql-sqlite/

Comment (SharkyKZ) [reply plugins/privacy/user/user.php]: So use `ParameterType::LARGE_OBJECT` here?

Comment (alikon) [reply plugins/privacy/user/user.php]: i didn't tested yet, but i think it should be the way to go

File: plugins/privacy/user/user.php [hunk 4]
```diff
@@ -134,20 +140,21 @@ public function onPrivacyRemoveData(PrivacyTableRequest $request, JUser $user =
...
 		$store     = JSessionStorage::getInstance($storeName);
-		$quotedIds = array();
+		$ids       = [];
 
 		// Destroy the sessions and quote the IDs to purge the session table
 		foreach ($sessionIds as $sessionId)
 		{
 			$store->destroy($sessionId);
-			$quotedIds[] = $this->db->quote($sessionId);
+			$ids[] = $sessionId;
```
Comment (SharkyKZ) [inline plugins/privacy/user/user.php]: Don't need `$ids` anymore. Use `$sessionIds`.

File: plugins/privacy/user/user.php (hunk 3 above)
Comment (SharkyKZ) [reply plugins/privacy/user/user.php]: Just tested it. We do need `ParameterType::LARGE_OBJECT` here, otherwise nothing is deleted.

Comment (SharkyKZ) [inline plugins/privacy/user/user.php]:
```suggestion
			->whereIn($db->quoteName('session_id'), $ids, ParameterType::LARGE_OBJECT);
```

**Comment by joeykrop** [comment]
I have tested this item :white_check_mark: successfully on ea60065034aa9fc1cc5e438356a846e422909e42<br /><br />Tested as much as possible.

Applying for reports, confirming and receiving also worked.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25432">issues.joomla.org/tracker/joomla-cms/25432</a>.</sub>

**Comment by crommie** [comment]
I have tested this item :white_check_mark: successfully on ea60065034aa9fc1cc5e438356a846e422909e42<br /><br />Played around with users. Works as expected.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25432">issues.joomla.org/tracker/joomla-cms/25432</a>.</sub>

File: plugins/privacy/user/user.php (hunk 3 above)
Comment (wilsonge) [reply plugins/privacy/user/user.php]: @HLeithner please can you fix this one please

Comment (HLeithner) [reply plugins/privacy/user/user.php]: thx done

Comment (HLeithner) [reply plugins/privacy/user/user.php]: Changed thx @wilsonge

**Comment by waader** [comment]
I have tested this item :white_check_mark: successfully on a4e188c23c81f73abea07eaa596771c76227a8f5<br /><br />Tested with postgresql.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25432">issues.joomla.org/tracker/joomla-cms/25432</a>.</sub>

**Comment by deepa-g** [comment]
I have tested this item :white_check_mark: successfully on 2ed0cbd07bba9de0d054a52aaabf7e16e368af2a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25432">issues.joomla.org/tracker/joomla-cms/25432</a>.</sub>

**Comment by vijaykhollam** [comment]
I have tested this item :white_check_mark: successfully on 2ed0cbd07bba9de0d054a52aaabf7e16e368af2a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25432">issues.joomla.org/tracker/joomla-cms/25432</a>.</sub>

**Comment by pravinTek** [comment]
I have tested this item :white_check_mark: successfully on 2ed0cbd07bba9de0d054a52aaabf7e16e368af2a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25432">issues.joomla.org/tracker/joomla-cms/25432</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25432">issues.joomla.org/tracker/joomla-cms/25432</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**